src/main/kotlin/exercise2/task4/ProcessCountriesData.kt

```diff
@@ -59,27 +59,33 @@ internal val countries = listOf(
  */
 
 internal fun List<Country>.findCountryWithBiggestTotalArea(): Country {
-    TODO("Implement me!!!")
+    return this.maxBy { it.totalAreaInSquareKilometers }
 }
 
 internal fun List<Country>.findCountryWithBiggestPopulation(): Country {
-    TODO("Implement me!!!")
+    return this.maxBy { it.population }
 }
 
 internal fun List<Country>.findCountryWithHighestPopulationDensity(): Country {
-    TODO("Implement me!!!")
+    return this.maxBy { it.population / it.totalAreaInSquareKilometers }
 }
 
 internal fun List<Country>.findCountryWithLowestPopulationDensity(): Country {
-    TODO("Implement me!!!")
+    return this.minBy { it.population / it.totalAreaInSquareKilometers }
 }
 
 internal fun List<Country>.findLanguageSpokenInMostCountries(): String {
-    TODO("Implement me!!!")
+    return this.flatMap { it.languages }
+        .groupingBy { it }
+        .eachCount()
+        .toList()
+        .sortedByDescending { it.second }
+        .map{ it -> it.first }
+        .first()
```

Comment:
Here, we can apply a small optimization and avoid unnecessary `map` calls that operate over the entire collection.  Since we are only interested in the single language that is used in most countries, instead of `map { ... }.first()` call, we can do the following: 
```suggestion
        .first()
        .first
```


```

Reply:
Also please note that `first()` function call will throw an exception if the receiver list object is empty.